libc/string_catdup.c

```diff
@@ -0,0 +1,802 @@
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(testdata_huge_str, buf, PATH_MAX - 1);
+	TEST_ASSERT_EQUAL_CHAR('\0', buf[PATH_MAX - 1]);
+}
+
+
+TEST(string_cat, strcat_specials_string)
+{
+	int i;
+	const char *expectedStr = SPECIALS_STRING SPECIALS_STRING;
+	char buf[SIZE_BUF * 4] = SPECIALS_STRING;
+	char buf1[] = SPECIALS_STRING;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(expectedStr, buf);
+
+	for (i = sizeof(SPECIALS_STRING SPECIALS_STRING) - 1; i < SIZE_BUF * 4; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_overwrite)
+{
+	int i;
+	char buf[SIZE_BUF];
+	char buf1[] = "orem ipsum";
+
+	memset(buf, 'X', sizeof(buf));
+	buf[0] = 'L';
+	buf[1] = '\0';
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+
+	TEST_ASSERT_EQUAL_CHAR('\0', buf[sizeof(WORD_1 WORD_2) - 1]);
+	for (i = sizeof(WORD_1 WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('X', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_ascii)
+{
+	int i;
+	char buf[SIZE_BIGGER_BUF] = { 0 };
+	char ascii[128] = { 0 };
+
+	for (i = 1; i < 128; i++) {
+		ascii[i - 1] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, ascii));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	for (i = sizeof(ascii) + 1; i < SIZE_BIGGER_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_extended_ascii)
+{
+	int i;
+	char buf[SIZE_BIGGER_BUF] = { 0 };
+	char ascii[129] = { 0 };
+
+	for (i = 128; i <= 255; i++) {
+		ascii[i - 128] = i;
+	}
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, ascii));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	for (i = sizeof(ascii) + 1; i < SIZE_BIGGER_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strcat_non_ascii)
+{
+	int i;
+	char buf[SIZE_BUF * 2] = NON_ASCII_1;
+	char buf1[] = NON_ASCII_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strcat(buf, buf1));
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, buf);
+
+	for (i = sizeof(NON_ASCII_1 NON_ASCII_2) + 1; i < SIZE_BUF * 2; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+///////////////////////////////////////////////////////////////////////////////////
+
+TEST(string_cat, strncat_basic)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, sizeof(WORD_2) + 1));
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+
+	for (i = sizeof(WORD_1 WORD_2); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_zero_n)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, 0));
+	TEST_ASSERT_EQUAL_STRING(WORD_1, buf);
+
+	for (i = sizeof(WORD_1); i < SIZE_BUF; i++) {
+		TEST_ASSERT_EQUAL_CHAR('\0', buf[i]);
+	}
+}
+
+
+TEST(string_cat, strncat_exceed)
+{
+	int i;
+	char buf[SIZE_BUF] = WORD_1;
+	char buf1[] = WORD_2;
+
+	TEST_ASSERT_EQUAL_PTR(buf, strncat(buf, buf1, sizeof(buf1) + 1));
```

Comment:
+1 not really valid here - if there woundn't be `\0` byte at the end of `buf1` you would try to inspect the `sizeof(buf1)+1` byte - so one byte outside of the `buf1`  (won't cause an error here as we know `\0` will be at the end of the string but still it's better to write valid code ;))